tools/src/wyvern/tools/parsing/coreparser/arch/ASTComponentTypeDecl.java

```diff
@@ -0,0 +1,175 @@
+        return provs;
+    }
+
+    public boolean isExternal() {
+        return isExternal;
+    }
+
+    public void setExternal(boolean isExternal) {
+        this.isExternal = isExternal;
+    }
+
+    public String getTypeName() {
+        return typeName;
+    }
+
+    public void setTypeName(String t) {
+        typeName = t;
+    }
+
+    public String toString() {
+        return super.toString() + " " + typeName + " : " + isExternal;
+    }
+
+    public void collectPorts() {
+        for (Node child : this.getChildren()) {
+            if (child instanceof ASTPortDecl) {
+                String req = ((ASTPortDecl) child).getRequires();
+                String prov = ((ASTPortDecl) child).getProvides();
+                String name = ((ASTPortDecl) child).getPort();
+                if (req != "") {
+                    reqs.put(name, req);
+                } else if (prov != "") {
+                    provs.put(name, prov);
+                }
+            }
+        }
+    }
+
+    public boolean checkModule(InterpreterState state) {
+        if (isExternal) {
+            return true;
+        }
+        Module mod = null;
+        int checkCount = 0;
+        try {
+            mod = state.getResolver().resolveModule(typeName);
+
+            /* Check that it's a module def */
+            Expression expr = mod.getExpression();
+            if (!expr.getType().toString().contains("apply(")) {
+                System.out.println(expr.getType().toString());
+                // throw error saying it's not a module def
+                ToolError.reportError(ErrorMessage.MODULE_DEF_NOT_FOUND, this.getLocation(),
+                        typeName);
+            }
+
+            /* Check for only for dependencies passed as args to module def */
+            if (expr instanceof New) {
+                for (Declaration d : ((New) expr).getDecls()) {
+                    if (d instanceof DefDeclaration) {
+                        List<FormalArg> args = ((DefDeclaration) d).getFormalArgs();
+                        if (args.size() != reqs.size()) {
+                            // inconsistent dependency errors
+                            ToolError.reportError(
+                                    ErrorMessage.COMPONENT_DEPENDENCY_INCONSISTENCY, this.getLocation(),
+                                    typeName);
+                        }
+                        for (FormalArg f : args) {
+                            String name = f.getName();
+                            String type = ((NominalType) f.getType()).getTypeMember();
+                            if (!reqs.get(name).equals(type)) {
+                                // inconsistent dependency error
+                                ToolError.reportError(
+                                        ErrorMessage.COMPONENT_DEPENDENCY_INCONSISTENCY, this.getLocation(),
+                                        typeName);
+                            }
+                        }
```

Comment:
Please check if the names match as well. This is required because you could have two ports with the same interface types. For matching names, you can assume that the arguments of the module def should be in the same order as the requires port declarations in the component type.